### DISCUSSION TITLE
How does socket() fail?

### DISCUSSION THREAD

**Discussion by orangepineapple** [ANLAB-KAIST/KENSv3#150] [Project Q&A 2. Transport Layer (TCP)] [answered]
I am testing the socket() function but I am a bit confused about how the socket call can fail. I minimized the socket  call to just the following code;
```
std::cout << "call" << std::endl;
int fd = createFileDescriptor(pid);    
std::cout << "pid: " << pid <<std::endl;
std::cout << "fd: " << fd << std::endl;
```

and I got an output like this:
call
pid: 0
fd: 3
 succ: 1 fail: 3 dup 0
/mnt/c/Users/brian/Comp_networks_KAIST/KENSv3-master/app/kens/testopen.cpp:59: Failure
Expected equality of these values:
  success
    Which is: 1
  ((size_t)test_size) * ((size_t)test_repeat)
    Which is: 4

/mnt/c/Users/brian/Comp_networks_KAIST/KENSv3-master/app/kens/testopen.cpp:60: Failure
Expected equality of these values:
  failed
    Which is: 3
  0

(I reduced the number of test cases)
I noticed that I was pretty much failing every call in the second loop of j but I have no idea why this would happen because my socket call is just called createFileDescriptor() and returning the fd

```
for (int j = 0; j < test_repeat; j++) {
      int fd_array[test_size];
      for (int k = 0; k < test_size; k++) {
        fd_array[k] = socket(AF_INET, SOCK_STREAM, IPPROTO_TCP);
        if (fd_array[k] < 0){
          failed++;
        }
        else{
          success++;
        }
      }

```

**Comment by orangepineapple** [comment]
I reduced the code further to get:
```
std::cout << "call" << std::endl;
this->returnSystemCall(syscallUUID, 1);
```
which I assumed would succeed on every one but just result in a bunch of duplicates, but I get and error like this:
call
test-kens-open: /mnt/c/Users/brian/Comp_networks_KAIST/KENSv3-master/src/Networking/E_Host.cpp:107: virtual E::Module::Message E::Host::messageReceived(E::ModuleID, E::Module::MessageBase&): Assertion `nsIter != appIter->second.fdToDomain.end()' failed.
Aborted

I am just wondering what this error means, and in detail how the SystemCalls work. I just assumed when the test code called socket it would call my implementation in the SystemCall and then the return value would be the value returned by returnSystemCall()

How can the SystemCall fail and return -1 without running any of my code and without me explicitly calling returnSystemCall( UUID , -1)?

**Comment by Pusnow** [comment]
You should return file descriptor numbers (returned from `createFileDescriptor `) using `returnSystemCall`. Your code just return 1, so it fails the test case.


> test-kens-open: /mnt/c/Users/brian/Comp_networks_KAIST/KENSv3-master/src/Networking/E_Host.cpp:107: virtual E::Module::Message E::Host::messageReceived(E::ModuleID, E::Module::MessageBase&): Assertion `nsIter != appIter->second.fdToDomain.end()' failed.
> I am just wondering what this error means, and in detail how the SystemCalls work. I just assumed when the test code called socket it would call my implementation in the SystemCall and then the return value would be the value returned by returnSystemCall()

Yes, when a system call is called, your `systemCallback` is called, and it returns when you call `returnSystemCall`.

The problem is that your implementation returns wrong file descriptors whose are not registered in KENS, and later test code uses the file descriptors, resulting in the assertion error.


> How can the SystemCall fail and return -1 without running any of my code and without me explicitly calling returnSystemCall( UUID , -1)?

At the end of the simulation, KENS forces all system calls to return -1.

**Reply by orangepineapple** [reply]
Sorry I think my question was not clear. So based on your answer, the following code should produce a correct test case:
```
std::cout << "call" << std::endl;
int fd = createFileDescriptor(pid);    
this->returnSystemCall(syscallUUID, fd);

```
 succ: 1024 fail: 130048 dup 66519552
/mnt/c/Users/brian/Comp_networks_KAIST/KENSv3-master/app/kens/testopen.cpp:59: Failure
Expected equality of these values:
  success
    Which is: 1024
  ((size_t)test_size) * ((size_t)test_repeat)
    Which is: 131072

/mnt/c/Users/brian/Comp_networks_KAIST/KENSv3-master/app/kens/testopen.cpp:60: Failure
Expected equality of these values:
  failed
    Which is: 130048
  0

/mnt/c/Users/brian/Comp_networks_KAIST/KENSv3-master/app/kens/testopen.cpp:61: Failure
Expected equality of these values:
  duplicated
    Which is: 66519552
  0
But when I run this code, it still fails the test case, and call is only printed 1024 times, the other test cases all return -1 and result in 130048 fails. I don't know why the call is returning -1.

**Reply by Pusnow** [reply, marked as answer]
Did you implement the close system call? You also need to use `removeFileDescriptor` to correctly manage file descriptors in KENS.

https://github.com/ANLAB-KAIST/KENSv3/wiki/Tip:-Managing-File-Descriptors

The 1024 successes mean your code works well on the first batch, but after `close()` is called, it seems your code do not works correctly.

> I don't know why the call is returning -1.

If you do not implement `close()`, your code is blocked when `close()` is called, pausing execution, and later all system calls (both `socket()` and `close()`) return -1 when simulation ends.

**Reply by orangepineapple** [reply]
OH! I did not (lol). I thought i wouldn't matter when unit testing the code. Thanks for the answer it works now!
